spotbugs/src/test/java/edu/umd/cs/findbugs/util/MutableClassesTest.java

```diff
@@ -54,6 +54,10 @@ public int getN() {
         public Immutable setN(int n) {
             return new Immutable(n);
         }
+
+        public void setNUnsupported(int n) throws UnsupportedOperationException {
```

Comment:
No, because we cannot check here the body of the function. IMO UnsupportedOperationException being thrown is better to be declared if the whole function call is not supported in this class.

Reply:
Can method declare to throw `UnsupportedOperationException` and modify state anyway? And even not throw anything at all?

Reply:
In the end, if code deliberately tries to mask bad behaviour, it will probably succeed.

Reply:
Yes, this is just a heuristics to find immutable classes. If you declare a function as a thrower but modify the state then this is a setter and the class is not immutable. The heuristics cannot be 100% accurate.

Reply:
Can it lead to false-negatives?

Reply:
Of course. As I mentioned, this is a heuristics. Unfortunately, unlike C++'s const member functions, Java has no means that forbids a method to modify the attributes of the class instance. Instead, we try to find methods which look like setters. If we do not find any then we consider the class for immutable. However, if the methods are named in a foreign language, or the setters are ill named (like eg. foo()), then we consider the class as immutable which leads to false negatives.

Reply:
I see, thank you  for your explanation.